client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/InternalFlag.java

```diff
@@ -0,0 +1,26 @@
+package org.infinispan.client.hotrod.impl.protocol;
+
+/**
+ * Flags which are internal to the protocol and shouldn't be exposed to users.
```

Comment:
@isavin The server can (and does) support multiple protocol versions, so I think non-backwards-compatible changes to the protocol are ok if they make the client simpler.

I agree that the user should only see the `put(k, v, duration, unit)` API and shouldn't worry about setting a flag. Actually, I think that also applies to `Flag.DEFAULT_LIFESPAN` and `Flag.DEFAULT_MAXIDLE` - users just call `cache.put(k, v)` and get the default values, and they shouldn't worry about setting some flag to get the default behaviour.

Now that I've remembered about `DEFAULT_LIFESPAN`, I'm leaning towards including the actual time unit in the protocol. We could encode it for both lifespan and maxIdle in a single byte, something like this:

```
public class ExpiryType {
   public static final byte SERVER_DEFAULTS = 0;
   public static final byte LIFESPAN_INFINITE = 1;
   public static final byte LIFESPAN_NANOS = 2;
   public static final byte LIFESPAN_MICROS = 3;
   public static final byte LIFESPAN_MILLIS = 4;
   public static final byte LIFESPAN_SECONDS = 5;
   public static final byte LIFESPAN_MINUTES = 6;
   public static final byte LIFESPAN_HOURS = 7;
   public static final byte LIFESPAN_DAYS = 8;
   public static final byte MAXIDLE_INFINITE = 0x10 + 1;
   public static final byte MAXIDLE_NANOS = 0x10 + 2;
   public static final byte MAXIDLE_MICROS = 0x10 + 3;
   public static final byte MAXIDLE_MILLIS = 0x10 + 4;
   public static final byte MAXIDLE_SECONDS = 0x10 + 5;
   public static final byte MAXIDLE_MINUTES = 0x10 + 6;
   public static final byte MAXIDLE_HOURS = 0x10 + 7;
   public static final byte MAXIDLE_DAYS = 0x10 + 8;
}
```

We could make the actual values optional, and only include them if the user specified a time unit. The `getWithMetadata` response is actually very similar, we should probably make them exactly the same.

There is one more problem: I don't think the specification for the lifespan and maxIdle parameters in the BasicCache API is clear enough. There is no mention of what happens if the values are 0, if maxIdle is negative, or if the lifespan is more than 30 days.

I would suggest something like this, although it probably can't make it in 7.2 for backwards compatibility reasons:
* No special cases if the `unit` parameter is non-null, except perhaps throwing an exception if the values are negative.
* Use the cache's defaults with `0/null` and no expiration with `-1/null` (or the other way around, since `0/null` is eternal in JCache). 
* Optional: Start throwing an exception for sub-millisecond time units, like JCache's `Duration` does.